crates/bevy_ecs/src/bundle.rs

```diff
@@ -174,6 +177,13 @@ pub unsafe trait Bundle: DynamicBundle + Send + Sync + 'static {
         // Ensure that the `OwningPtr` is used correctly
         F: for<'a> FnMut(&'a mut T) -> OwningPtr<'a>,
         Self: Sized;
+
+    /// Registers components that are required by the components in this [`Bundle`].
+    fn register_required_components(
+        _components: &mut Components,
+        _storages: &mut Storages,
+        _required_components: &mut RequiredComponents,
```

Comment:
The fact that this reference needs to be mutable is surprising to me.

Reply:
`_components` and `_storages` are `mut` because we need to be able to initialize components that haven't been initialized yet (to retrieve the ComponentId).

`_required_components`  is `mut` because it is the collection of components, which we are adding to in this function. This is where the "registration" happens.